packages/actor-query-operation-from/lib/ActorQueryOperationFrom.ts

```diff
@@ -0,0 +1,122 @@
+    const copiedOperation: Algebra.Operation = <any> {};
+    for (const key of Object.keys(operation)) {
+      if (Array.isArray(operation[key])) {
+        if (key === 'variables') {
+          copiedOperation[key] = operation[key];
+        } else {
+          copiedOperation[key] = operation[key]
+            .map((subOperation: any) => this.applyOperationGraph(subOperation, graph));
+        }
+      } else if (ActorQueryOperationFrom.ALGEBRA_TYPES.indexOf(operation[key].type) >= 0) {
+        copiedOperation[key] = this.applyOperationGraph(operation[key], graph);
+      } else {
+        copiedOperation[key] = operation[key];
+      }
+    }
+    return copiedOperation;
+  }
+
+  /**
+   * Transform the given array of operations into a union operation.
+   * @param {Operation[]} operations An array of operations, must contain at least two operations.
+   * @return {Union} A union operation.
+   */
+  public static unionOperations(operations: Algebra.Operation[]): Algebra.Union {
+    if (operations.length === 2) {
+      return ActorQueryOperationFrom.FACTORY.createUnion(operations[0], operations[1]);
+    } else if (operations.length > 2) {
+      return ActorQueryOperationFrom.FACTORY.createUnion(operations.shift(), this.unionOperations(operations));
+    } else {
+      throw new Error('A union can only be applied on at least two operations');
+    }
+  }
+
+  /**
+   * Transform a context based on the named graphs in the pattern.
+   * FROM NAMED indicates which named graphs are available.
+   * Add these graphs to the context so that quad pattern resolvers can use this information
+   * to ONLY select data from these graphs.
+   * @param {From} pattern
+   * @param {{[p: string]: any}} context A context.
+   * @return {{[p: string]: any}} A transformed context.
+   */
+  public static transformContext(pattern: Algebra.From, context?: { [id: string]: any }): { [id: string]: any } {
+    if (pattern.named.length) {
+      context = Object.assign({}, context || {});
+      if (context.namedGraphs) {
+        context.namedGraphs = require('lodash.intersectionwith')(context.namedGraphs, pattern.named,
+          (ths: RDF.Term, tht: RDF.Term) => ths.equals(tht));
+      } else {
+        context.namedGraphs = pattern.named;
+      }
+    }
+    return context;
+  }
+
+  /**
+   * Transform an operation based on the default graphs in the pattern.
+   * FROM sets the default graph.
+   * If multiple are available, take the union of the operation for all of them.
+   * @param {From} pattern A from operation.
+   * @return {Operation} The transformed operation.
+   */
+  public static createOperation(pattern: Algebra.From): Algebra.Operation {
+    if (!pattern.default.length) {
+      return pattern.input;
+    } else if (pattern.default.length === 1) {
+      return ActorQueryOperationFrom.applyOperationGraph(pattern.input, pattern.default[0]);
+    } else {
+      return ActorQueryOperationFrom.unionOperations(pattern.default.map(
```

Comment:
You're right, I didn't think of it that way. This means that the QPF-enabled ldf client also did this wrong...

I'll change the implementation so that the union is pushed down to quad pattern-level. That should lead to correct results as far as I can see.